File: test/CSharp/CodeCracker.Test/Style/InterfaceNameTests.cs
using CodeCracker.CSharp.Style;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Testing;
using System.Threading.Tasks;
using Xunit;

namespace CodeCracker.Test.CSharp.Style
{
    public class InterfaceNameTests : CodeFixVerifier<InterfaceNameAnalyzer, InterfaceNameCodeFixProvider>
    {
        [Fact]
        public async Task InterfaceNameStartsWithLetterI()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        public interface IFoo
        {
            void Test();
        }
    }";
            await VerifyCSharpHasNoDiagnosticsAsync(source);
        }

        [Fact]
        public async Task InterfaceNameNotStartsWithLetterI()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        public interface Foo
        {
            void Test();
        }
    }";
            var expected = new DiagnosticResult(DiagnosticId.InterfaceName.ToDiagnosticId(), DiagnosticSeverity.Info)
                .WithLocation(4, 9)
                .WithMessage(InterfaceNameAnalyzer.MessageFormat);
            await VerifyCSharpDiagnosticAsync(source, expected);
        }

        [Fact]
        public async Task ChangeInterfaceNameWithoutI()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        public interface Foo
        {
            void Test();
        }
    }";
            const string fixtest = @"
    namespace ConsoleApplication1
    {
        public interface IFoo
        {
            void Test();
        }
    }";
            await VerifyCSharpFixAsync(source, fixtest);
        }


        [Fact]
        public async Task ChangeInterfaceNameWithoutIAndClassImplementation()
        {
            const string source = @"
    namespace ConsoleApplication1
    {
        public interface Foo
        {
            void Test();
        }

        public class Test : Foo
        {
            public void Test()
            {

            }
        }
    }";
            const string fixtest = @"
    namespace ConsoleApplication1
    {
        public interface IFoo
        {
            void Test();
        }

        public class Test : IFoo
        {
            public void Test()
            {

            }
        }
    }";
            await VerifyCSharpFixAsync(source, fixtest);
        }

        [Fact]
        public async Task ChangeAllInterfaceNamesWithoutI()
        {
            const string source1 = @"
    namespace ConsoleApplication1
    {
        public interface Foo1
        {
            void Test();
        }
    }";
            const string source2 = @"
    namespace ConsoleApplication2
    {
        public interface Foo2
        {
            void Test();
        }
    }";
            const string fixtest1 = @"
    namespace ConsoleApplication1
    {
        public interface IFoo1
        {
            void Test();
        }
    }";
            const string fixtest2 = @"
    namespace ConsoleApplication2
    {
        public interface IFoo2
        {
            void Test();
        }
    }";
            await VerifyCSharpFixAllAsync(new string[] { source1, source2 }, new string[] { fixtest1, fixtest2 });
        }


        [Fact]
        public async Task ChangeAllInterfaceNamesWithoutIAndClassImplementation()
        {
            const string source1 = @"
    using ConsoleApplication2;
    namespace ConsoleApplication1
    {
        public interface Foo2
        {
            void Test();
        }

        public class Test1 : Foo1
        {
            public void Test()
            {

            }
        }
    }";
            const string source2 = @"
    using ConsoleApplication1;
    namespace ConsoleApplication2
    {
        public interface Foo1
        {
            void Test();
        }

        public class Test2 : Foo2
        {
            public void Test()
            {

            }
        }
    }";
            const string fixtest1 = @"
    using ConsoleApplication2;
    namespace ConsoleApplication1
    {
        public interface IFoo2
        {
            void Test();
        }

        public class Test1 : IFoo1
        {
            public void Test()
            {

            }
        }
    }";
            const string fixtest2 = @"
    using ConsoleApplication1;
    namespace ConsoleApplication2
    {
        public interface IFoo1
        {
            void Test();
        }

        public class Test2 : IFoo2
        {
            public void Test()
            {

            }
        }
    }";
            await VerifyCSharpFixAllAsync(new string[] { source1, source2 }, new string[] { fixtest1, fixtest2 });
        }
    }
}

File: test/VisualBasic/CodeCracker.Test/Style/InterfaceNameTests.vb
Imports CodeCracker.VisualBasic.Style
Imports Microsoft.CodeAnalysis.Testing
Imports Xunit

Namespace Style
    Public Class InterfaceNameTests
        Inherits CodeFixVerifier(Of InterfaceNameAnalyzer, InterfaceNameCodeFixProvider)
        <Fact>
        Public Async Function InterfaceNameStartsWithLetterI() As Task
            Const source = "Namespace ConsoleApplication1
    Public Interface IFoo
        Sub Test()
    End Interface
End Namespace"

            Await VerifyBasicHasNoDiagnosticsAsync(source)
        End Function

        <Fact>
        Public Async Function InterfaceNameNotStartsWithLetterI() As Task
            Const source = "Namespace ConsoleApplication1
    Public Interface Foo
        Sub Test()
    End Interface
End Namespace"

            Dim expected = New DiagnosticResult(DiagnosticId.InterfaceName.ToDiagnosticId(), Microsoft.CodeAnalysis.DiagnosticSeverity.Info) _
                .WithLocation(2, 5) _
                .WithMessage(InterfaceNameAnalyzer.MessageFormat)

            Await VerifyBasicDiagnosticAsync(source, expected)
        End Function

        <Fact>
        Public Async Function ChangeInterfaceNameWithoutI() As Task
            Const source = "Namespace ConsoleApplication1
    Public Interface Foo
        Sub Test()
    End Interface
End Namespace"
            Const fix = "Namespace ConsoleApplication1
    Public Interface IFoo
        Sub Test()
    End Interface
End Namespace"
            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function ChangeInterfaceNameWithoutIAndClassImplementation() As Task
            Const source = "Namespace ConsoleApplication1
    Public Interface Foo
        Sub Test()
    End Interface
    Public Class Test
        Implements Foo
        Public Sub Test() Implements Foo.Test
        End Sub
    End Class
End Namespace"
            Const fix = "Namespace ConsoleApplication1
    Public Interface IFoo
        Sub Test()
    End Interface
    Public Class Test
        Implements IFoo
        Public Sub Test() Implements IFoo.Test
        End Sub
    End Class
End Namespace"

            Await VerifyBasicFixAsync(source, fix)
        End Function

        <Fact>
        Public Async Function ChangeAllInterfaceNamesWithoutI() As Task
            Const source1 = "Namespace ConsoleApplication1
    Public Interface Foo1
        Sub Test()
    End Interface
End Namespace"
            Const source2 = "Namespace ConsoleApplication2
    Public Interface Foo2
        Sub Test()
    End Interface
End Namespace"

            Const fix1 = "Namespace ConsoleApplication1
    Public Interface IFoo1
        Sub Test()
    End Interface
End Namespace"
            Const fix2 = "Namespace ConsoleApplication2
    Public Interface IFoo2
        Sub Test()
    End Interface
End Namespace"

            Await VerifyBasicFixAllAsync(New String() {source1, source2}, New String() {fix1, fix2})
        End Function

        <Fact>
        Public Async Function ChangeAllInterfaceNamesWithoutIAndClassImplementation() As Task
            Const source1 = "Namespace ConsoleApplication1
    Public Interface Foo1
        Sub Test()
    End Interface

    Public Class Test1
        Implements Foo1
        Public Sub Test() Implements Foo1.Test
        End Sub
    End Class
End Namespace"
            Const source2 = "Namespace ConsoleApplication2
    Public Interface Foo2
        Sub Test()
    End Interface

    Public Class Test2
        Implements Foo2
        Public Sub Test() Implements Foo2.Test
        End Sub
    End Class
End Namespace"

            Const fix1 = "Namespace ConsoleApplication1
    Public Interface IFoo1
        Sub Test()
    End Interface

    Public Class Test1
        Implements IFoo1
        Public Sub Test() Implements IFoo1.Test
        End Sub
    End Class
End Namespace"
            Const fix2 = "Namespace ConsoleApplication2
    Public Interface IFoo2
        Sub Test()
    End Interface

    Public Class Test2
        Implements IFoo2
        Public Sub Test() Implements IFoo2.Test
        End Sub
    End Class
End Namespace"

            Await VerifyBasicFixAllAsync(New String() {source1, source2}, New String() {fix1, fix2})
        End Function

    End Class
End Namespace